File: LeetCode/0515. Find Largest Value in Each Tree Row/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

/**
 * 78 / 78 test cases passed.
 * Runtime: 4 ms
 * Memory Usage: 21.6 MB 
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        vector<int> ans;
        if (root == nullptr) return ans;
        queue<TreeNode*> q;
        q.emplace(root);
        while (!q.empty()) {
            int size = q.size();
            int level_max = INT_MIN;
            for (int i = 0; i < size; ++ i) {
                auto node = q.front(); q.pop();
                level_max = max(level_max, node->val);
                if (node->right) q.emplace(node->right);
                if (node->left) q.emplace(node->left);
            }
            ans.emplace_back(level_max);
        }
        return ans;
    }
};

File: LeetCode/0515. Find Largest Value in Each Tree Row/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

"""
78 / 78 test cases passed.
Runtime: 52 ms
Memory Usage: 17.6 MB
"""
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        if not root:
            return ans
        q = collections.deque([root])
        while q:
            level_max = float('-inf')
            for _ in range(len(q)):
                n = q.popleft()
                level_max = max(level_max, n.val)
                if n.left:
                    q.append(n.left)
                if n.right:
                    q.append(n.right)
            ans.append(level_max)
        return ans
